### PR TITLE
Add custom token management

### LINKED ISSUE #1160: Add custom token
### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

- Display only certified tokens by default
- Add a feature to add a custom token in local (not onchain)
- Display tokens list (certified + custom tokens)
For each account

### Additional context

_No response_

### Epic

_No response_

### PR DESCRIPTION
**Pull request by redDwarf03**
# Description

Fixes #1160

## Type of change

- New feature (non-breaking change which adds functionality)

## Material used:

Please delete options that are not relevant.
- iOS (Smartphone/Tablet)

## How Has This Been Tested?

- Add / Hide custom tokens
- In add custom sheet, search token by address or directly in the user's balance
- In swap sheet, search token by address or in custom list

## Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/application/tokens/tokens.dart [hunk 1]
```diff
@@ -1,49 +1,81 @@
...
-Future<List<aedappfm.AEToken>> tokensList(
-  TokensListRef ref,
-  String userGenesisAddress, {
+Future<Map<String, archethic.Token>> tokensFromAddresses(
+  Ref ref,
+  List<String> addresses,
+) async {
+  final apiService = ref.watch(apiServiceProvider);
+
+  return TokensRepositoryImpl().getTokensFromAddresses(
```
Comment (Chralu) [inline lib/application/tokens/tokens.dart]: Ce serait pas mal d'avoir un provider pour faire un Singleton de ce repository.

File: lib/domain/models/token_parser.dart [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if (isLPToken) {
+      final poolRaw = poolsListRaw.firstWhereOrNull(
+        (item) => item.lpTokenAddress == token.address!,
+      );
+      if (poolRaw != null) {
+        token1Address =
+            poolRaw.concatenatedTokensAddresses.split('/')[0].toUpperCase();
+        token2Address =
+            poolRaw.concatenatedTokensAddresses.split('/')[1].toUpperCase();
+        if (token1Address != 'UCO') {
```
Comment (Chralu) [inline lib/domain/models/token_parser.dart]: Ideally, we should use a constant for `'UCO'` value.

File: lib/domain/models/token_parser.dart [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import 'package:aewallet/infrastructure/repositories/tokens/tokens.repository.dart';
+import 'package:aewallet/modules/aeswap/domain/models/util/get_pool_list_response.dart';
+import 'package:archethic_dapp_framework_flutter/archethic_dapp_framework_flutter.dart'
+    as aedappfm;
+import 'package:archethic_lib_dart/archethic_lib_dart.dart' as archethic;
+import 'package:collection/collection.dart';
+
+mixin TokenParser {
+  Future<aedappfm.AEToken> tokenModelToAETokenModel(
```
Comment (Chralu) [inline lib/domain/models/token_parser.dart:10]: Regarding to this function complexity, some automated tests would be useful.

File: lib/infrastructure/repositories/tokens/tokens.repository.dart [hunk 4]
```diff
@@ -47,37 +46,40 @@ class TokensRepositoryImpl implements TokensRepository {
 
     final getTokens = await Future.wait(futures);
     for (final Map<String, archethic.Token> getToken in getTokens) {
-      tokenMap.addAll(getToken);
-
       getToken.forEach((key, value) async {
-        value = value.copyWith(address: key);
-        await tokensListDatasource.setToken(value.toHive());
+        if (value.type == 'fungible') {
```
Comment (Chralu) [inline lib/infrastructure/repositories/tokens/tokens.repository.dart]: String magique détecté.

Pour corriger, par exemple, ajouter une méthode `bool get isFungible` à `archethic.Token`.

File: lib/ui/views/add_custom_token/bloc/provider.dart [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      final tokenResult =
+          await ref.read(tokensFromAddressesProvider([tokenAddress]).future);
+      if (tokenResult[tokenAddress] != null) {
+        final userTokenBalance = await ref.read(
+          getBalanceProvider(
+            tokenAddress,
+          ).future,
+        );
+        final poolsListRaw =
+            await ref.watch(DexPoolProviders.getPoolListRaw.future);
```
Comment (Chralu) [inline lib/ui/views/add_custom_token/bloc/provider.dart]: Je pense qu'il faudrait un `read` ici.

File: lib/ui/views/add_custom_token/bloc/provider.dart [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        );
+      } else {
+        setError(
+          appLocalizations.customTokenAddressNotValid,
+        );
+      }
+    }
+  }
+
+  void setError(
```
Comment (Chralu) [inline lib/ui/views/add_custom_token/bloc/provider.dart:79]: Shouldn't this be private ?

File: lib/ui/views/add_custom_token/layouts/add_custom_token_sheet.dart [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                AppButtonTinyConnectivity(
+                  localizations.confirm,
+                  disabled: !addCustomTokenForm.isControlsOk,
+                  Dimens.buttonBottomDimens,
+                  key: const Key('confirm'),
+                  onPressed: () async {
+                    final result = await ref
+                        .read(addCustomTokenFormNotifierProvider.notifier)
+                        .addCustomToken(ref, localizations);
+                    if (result) {
```
Comment (Chralu) [inline lib/ui/views/add_custom_token/layouts/add_custom_token_sheet.dart]: Plutot faire un `if (!result) return;`

File: lib/ui/views/add_custom_token/bloc/provider.dart (hunk 6 above)
Comment (redDwarf03) [reply lib/ui/views/add_custom_token/bloc/provider.dart:79]: No. it's used in `add_custom_token_sheet.dart`

File: lib/infrastructure/repositories/tokens/tokens.repository.dart (hunk 4 above)
Comment (redDwarf03) [reply lib/infrastructure/repositories/tokens/tokens.repository.dart]: I use a constant but not in libdart for the moment

**Review by Chralu — APPROVED** [review]

**Pull request merged**